pkg/storage/ingest/reader_test.go

```diff
@@ -843,6 +843,84 @@ func TestPartitionReader_ConsumeAtStartup(t *testing.T) {
 			return reader.State()
 		})
 	})
+
+	t.Run("should not wait indefinitely if there are no records to consume from Kafka but partition start offset is > 0 (e.g. all previous records have been deleted by Kafka retention)", func(t *testing.T) {
+		t.Parallel()
+
+		for _, consumeFromPosition := range consumeFromPositionOptions {
+			consumeFromPosition := consumeFromPosition
+
+			t.Run(fmt.Sprintf("consume from position: %s", consumeFromPosition), func(t *testing.T) {
+				t.Parallel()
+
+				ctx, cancel := context.WithCancel(context.Background())
+				t.Cleanup(cancel)
+
+				consumer := consumerFunc(func(ctx context.Context, records []record) error {
+					return nil
+				})
+
+				cluster, clusterAddr := testkafka.CreateCluster(t, partitionID+1, topicName)
+				cluster.ControlKey(int16(kmsg.Fetch), func(req kmsg.Request) (kmsg.Response, error, bool) {
+					cluster.KeepControl()
+
+					// Throttle the Fetch request.
+					select {
+					case <-ctx.Done():
+					case <-time.After(time.Second):
+					}
```

Comment:
FWIW, if you're interested, I've since introduced this function:

https://pkg.go.dev/github.com/twmb/franz-go/pkg/kfake#Cluster.SleepControl

which allows the rest of the kfake cluster to run _other_ requests that are on _different_ connections, while hanging the connection responsible for this actively running request.

Reply:
Cool, nice, thanks! I love you helping us with your library 🤗